In [0]:
from google.colab import drive
drive.mount('/content/drive')
basedir = '/content/drive/My Drive/PSDA Group 4/Data/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
from sklearn import preprocessing
import pandas as pd
import numpy as np
import csv
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import RFECV

In [0]:
data_path = basedir + 'wdbc.data'

In [0]:
data = pd.read_csv(data_path,header=None)
data #check the rows and columns of this dataset

0  1      2      3   ...      28      29      30       31
0      842302  M  17.99  10.38  ...  0.7119  0.2654  0.4601  0.11890
1      842517  M  20.57  17.77  ...  0.2416  0.1860  0.2750  0.08902
2    84300903  M  19.69  21.25  ...  0.4504  0.2430  0.3613  0.08758
3    84348301  M  11.42  20.38  ...  0.6869  0.2575  0.6638  0.17300
4    84358402  M  20.29  14.34  ...  0.4000  0.1625  0.2364  0.07678
..        ... ..    ...    ...  ...     ...     ...     ...      ...
564    926424  M  21.56  22.39  ...  0.4107  0.2216  0.2060  0.07115
565    926682  M  20.13  28.25  ...  0.3215  0.1628  0.2572  0.06637
566    926954  M  16.60  28.08  ...  0.3403  0.1418  0.2218  0.07820
567    927241  M  20.60  29.33  ...  0.9387  0.2650  0.4087  0.12400
568     92751  B   7.76  24.54  ...  0.0000  0.0000  0.2871  0.07039

[569 rows x 32 columns]

In [0]:
x = np.zeros((569, 1), dtype=float) #set array to store the serial number
y = np.zeros((569, 30), dtype=float) #set array to store the values
z = np.zeros((569,), dtype=str)# #set array to store the labels

In [0]:
#read the dataset into array
with open(data_path, 'r') as input_file:
    reader = csv.reader(input_file)
    for i, row in enumerate(reader):
        data_name = [float(row[0])] 
        x[i] = data_name
        data_value = [float(datum) for datum in row[2:] ]
        y[i] = data_value  
        z[i] = row[1] #label
#     y = np.concatenate((x, y), axis=1)

In [0]:
#Encode the labels with value 1 and  0
le = preprocessing.LabelEncoder()
le.fit(["M", "B"])  #Label: M(M=malignant) is 1, B(B=benign) is 0.
z = le.transform(z)
# print('label standard:%s' % data_label)
# print('reverse the label:%s' % le.inverse_transform([0, 0, 1]))

In [0]:
#Split the dataset into training set and test set
y_train, y_test, z_train, z_test = train_test_split(y, z, test_size = 0.2, random_state = 1)

#Transformer and Estimator step by step

In [0]:
#=============================================================

Feature scaling

In [0]:
ytrain_transformed = StandardScaler().fit_transform(y_train)
ytest_transformed = StandardScaler().fit_transform(y_test)

Use PCA to reduce the dimensionality

In [0]:
pca = PCA(n_components=2)
y_train_pca = pca.fit_transform(ytrain_transformed)
y_test_pca = pca.fit_transform(ytest_transformed)


Classifier

In [0]:
classifier = LogisticRegression(random_state=1, solver='liblinear')
classifier.fit(y_train_pca, z_train) #z_train is label
print('Test accuracy: %.3f' % classifier.score(y_test_pca, z_test))

Test accuracy: 0.789


In [0]:
#=============================================================

#Use pipeline to pack all steps(Transformer, Estimator)

---



In [0]:
pipeline = Pipeline([
    ('sc', StandardScaler()),
    ('pca', PCA(n_components=2)),
    ('clf', LogisticRegression(random_state=1, solver='liblinear'))
])
pipeline.fit(y_train, z_train)
print('Test accuracy: %.3f' % pipeline.score(y_test, z_test))

Test accuracy: 0.947


# Recursive Feature Elimination

In [0]:
selector = RFECV(LogisticRegression(random_state=1, solver='liblinear'), step=1, cv=5)
selector = selector.fit(y, z)
print(selector.support_)
print(selector.ranking_)

[ True  True  True False  True  True  True  True  True  True  True  True
  True  True False False  True  True  True False  True  True  True  True
  True  True  True  True  True  True]
[1 1 1 5 1 1 1 1 1 1 1 1 1 1 2 3 1 1 1 4 1 1 1 1 1 1 1 1 1 1]


We see that 26 of the 30 features are relevant. The following features (according to their index number) are relevant and hence used.

In [0]:
y_train_rfe = y_train[:,[0,1,2,4,5,6,7,8,9,10,11,12,13,16,17,18,20,21,22,23,24,24,26,27,28,29]]
y_test_rfe = y_test[:,[0,1,2,4,5,6,7,8,9,10,11,12,13,16,17,18,20,21,22,23,24,24,26,27,28,29]]

In [0]:
pipeline = Pipeline([
    ('sc', StandardScaler()),
    ('clf', LogisticRegression(random_state=1, solver='liblinear'))
])
pipeline.fit(y_train_rfe, z_train)
print('Test accuracy: %.3f' % pipeline.score(y_test_rfe, z_test))

Test accuracy: 0.974


With the Feature Elimination instead of the PCA we improve our accuracy score to 97.4%

#Conclusion

In a typical machine learning workflow you will need to apply all these transformations at least twice. Once when training the model and again on any new data you want to predict on. Of course you could write a function to apply them and reuse that but you would still need to run this first and then call the model separately. Scikit-learn pipelines are a tool to simplify this process. They have several key benefits:
*   They make your workflow much easier to read and understand.
*   They enforce the implementation and order of steps in your project.
*   These in turn make your work much more reproducible.

Scikit-learn has built in functions for most of these commonly used transformations in it’s package.



